oxidation/libparsec/crates/testbed/src/env.rs

```diff
@@ -42,6 +51,38 @@ pub struct TestbedEnv {
     // given config dir, at which point they should use `test_get_testbed` and
     // initialize `persistence_store` by using the data from `template`.
     pub persistence_store: Mutex<HashMap<&'static str, Box<dyn Any + Send>>>,
+    // Must keep track of the generated local devices to avoid nasty behavior
+    // if the test call generate the local device multiple times for the same device.
+    local_devices_cache: Mutex<Vec<Arc<LocalDevice>>>,
+}
+
+impl TestbedEnv {
+    pub fn local_device(&self, device_id: DeviceID) -> Arc<LocalDevice> {
+        let mut guard = self.local_devices_cache.lock().expect("Mutex is poisoned");
+        for item in guard.iter() {
+            if item.device_id == device_id {
+                return item.clone();
+            }
+        }
+        // Not found, must generate it
+        let device = self.template.device(&device_id);
+        let user = self.template.user(device_id.user_id());
+        let local_device = Arc::new(LocalDevice {
+            organization_addr: self.organization_addr.to_owned(),
+            device_id,
+            device_label: device.device_label.to_owned(),
+            human_handle: user.human_handle.to_owned(),
+            signing_key: device.signing_key.to_owned(),
+            private_key: user.private_key.to_owned(),
+            profile: user.profile.to_owned(),
+            user_manifest_id: user.user_manifest_id.to_owned(),
+            user_manifest_key: user.user_manifest_key.to_owned(),
+            local_symkey: device.local_symkey.to_owned(),
+            time_provider: TimeProvider::default(),
+        });
```

Comment:
Maybe juste clone `user` and `device` ?

Reply:
I don't see the gain for this, multiple `to_owned()` is more verbose, but that avoid to clone useless data (there is not much for the moment, but `TestbedUserData`/`TestbedDeviceData` may contain more stuff in the future with the support of manifest/blocks in the testbed...)